### ISSUE TITLE
A way to delete mapping with specific lhs.

### ISSUE THREAD

**Issue by hoofcushion** [neovim/neovim#30258]
### Problem

`unmap` or `nvim_del_keymap` will delete all mapping with matched lhs or rhs, there should be a way to only delete mapping with matched lhs.

For example:
```lua
vim.api.nvim_set_keymap("n","ge","q",{})
vim.api.nvim_del_keymap("n","q")
print(vim.api.nvim_command("verbose map ge")) -- No mapping found
```

nvim_del_keymap will delete mapping `normal: ge -> q`  even its lhs doesn't match.

There is no way to keep other mapping has `q` as rhs after unmap or nvim_del_keymap `q`.

### Expected behavior

A api to delete mapping with matched lhs only.

**Comment by justinmk** [comment]
If we implement this, all the old `nvim_x_keymap` APIs should be deprecated and we should introduce `nvim_keymap_x` instead.

**Comment by justinmk** [comment]
I had no idea that `:unmap` works like that, but indeed it appears to. wtf?

```
:nnoremap a b
:map a  " prints mapping
:nunmap b
:map a  " nothing
```

**Comment by clason** [comment]
That's a regression in Vim, apparently.

**Comment by hoofcushion** [comment]
Well, i did check the doc for this behavior, It was a feature, not a bug.

**Comment by clason** [comment]
Well, we have a choice here: Do we want the API to match the Vimscript behavior or not? If we don't (now), this is a breaking change (which the API contract doesn't allow). If we do, you'll have to get Vim to change this first.

If you're fine with such a change being made to the Lua API (`vim.keymap.del`), we have more leeway.

**Comment by justinmk** [comment]
It could be an optional `lhs:boolean` flag, though that would require a new API. 

I doubt any API consumer is intentionally using this feature, so we could consider unconditionally having "LHS only" behavior for `nvim_del_keymap`.

**Comment by hoofcushion** [comment]
I think neovim deserve a better implementation if the old one wasn't good, even it breaks the capabilities from vim to neovim.

**Comment by clason** [comment]
> It could be an optional `lhs:boolean` flag, though that would require a new API. 
> 
> 
> 
> I doubt any API consumer is intentionally using this feature, so we could consider unconditionally having "LHS only" behavior for `nvim_del_keymap`.

Yes, I believe so. The documentation is clear enough that this would count as "fix". There's still vim.cmd.unmap if someone wants the vim way.

**Comment by hoofcushion** [comment]
Can nvim_set_keymap return a id? Get id by lhs then delete mapping by id is more clear, just like nvim_del_autocmd.

**Comment by zeertzjq** [comment]
The API contract doesn't allow that.

**Comment by justinmk** [comment]
> The API contract doesn't allow such modifications to existing APIs.

I think we should try relaxing the API contract for changing `void => non-void` return type. ( ✅ Done: https://github.com/neovim/neovim/pull/34811 )

**Comment by hoofcushion** [comment]
Well, i did check the doc for this behavior, It was a feature, not a bug.

**Comment by hoofcushion** [comment]
I had a work-around for this problem: 
Connect "<ignore>" after the rhs when creating keymap. Vim will not be able to match it.
